go/nbs/s3_table_persister.go

```diff
@@ -184,3 +198,233 @@ func (s partsByPartNum) Less(i, j int) bool {
+		select {
+		case sentPart, open := <-sent:
+			if open {
+				multipartUpload.Parts = append(multipartUpload.Parts, &s3.CompletedPart{
+					ETag:       aws.String(sentPart.etag),
+					PartNumber: aws.Int64(sentPart.idx),
+				})
+			}
+			cont = open
+
+		case err := <-failed:
+			if err != nil && firstFailure == nil { // nil err may happen when failed gets closed
+				firstFailure = err
+				close(done)
+			}
+		}
+	}
+
+	// If there was any failure detected above, |done| is already closed
+	if firstFailure == nil {
+		close(done)
+	}
+	sort.Sort(partsByPartNum(multipartUpload.Parts)) // S3 requires that these be in part-order
+	return multipartUpload, firstFailure
+}
+
+type copyPart struct {
+	name              string
+	srcOffset, srcLen int64
+}
+
+type manualPart struct {
+	srcR             io.Reader
+	dstStart, dstEnd int64
+}
+
+func dividePlan(plan compactionPlan, minPartSize, maxPartSize uint64) (copies []copyPart, manuals []manualPart, buff []byte) {
+	// NB: if maxPartSize < 2*minPartSize, splitting large copies apart isn't solvable. S3's limits are plenty far enough apart that this isn't a problem in production, but we could violate this in tests.
+	d.PanicIfTrue(maxPartSize < 2*minPartSize)
+
+	buffSize := uint64(len(plan.mergedIndex))
+	i := 0
+	for ; i < len(plan.sources); i++ {
+		sws := plan.sources[i]
+		if sws.dataLen < minPartSize {
+			// since plan.sources is sorted in descending chunk-data-length order, we know that sws and all members after it are too small to copy.
+			break
+		}
+		if sws.dataLen <= maxPartSize {
+			copies = append(copies, copyPart{sws.source.hash().String(), 0, int64(sws.dataLen)})
+			continue
+		}
+
+		// Now, we need to break the data into some number of parts such that for all parts minPartSize <= size(part) <= maxPartSize. This code tries to split the part evenly, such that all new parts satisfy the previous inequality. This gets tricky around edge cases. Consider min = 5b and max = 10b and a data length of 101b. You need to send 11 parts, but you can't just send 10 parts of 10 bytes and 1 part of 1 byte -- the last is too small. You also can't send 10 parts of 9 bytes each and 1 part of 11 bytes, because the last is too big. You have to distribute the extra bytes across all the parts so that all of them fall into the proper size range.
+		lens := calcPartLens(sws.dataLen, maxPartSize)
+
+		var srcStart int64
+		for _, length := range lens {
+			copies = append(copies, copyPart{sws.source.hash().String(), srcStart, length})
+			srcStart += length
+		}
+	}
+	var offset int64
+	for ; i < len(plan.sources); i++ {
+		sws := plan.sources[i]
+		manuals = append(manuals, manualPart{sws.source.reader(), offset, offset + int64(sws.dataLen)})
+		offset += int64(sws.dataLen)
+		buffSize += sws.dataLen
+	}
+	buff = make([]byte, buffSize)
+	copy(buff[buffSize-uint64(len(plan.mergedIndex)):], plan.mergedIndex)
+	return
+}
+
+func calcPartLens(dataLen, maxPartSize uint64) []int64 {
```

Comment:
suggest `calcPartLens` => `splitOnMaxSize`.

Add a note here describing what this does. `splits 'dataLen' into (the maximum) N roughly equal (within 1 byte) lengths such that each is <= maxPartSize`